lib/pure/bitopts.nim

```diff
@@ -0,0 +1,283 @@
+            57, 46, 52, 38, 26, 32, 41, 50, 36, 17, 19, 29, 10, 13, 21, 56, 45, 25, 31,
+            35, 16, 9, 12, 44, 24, 15, 8, 23, 7, 6, 5, 63]
+        var v = x.uint64
+        v = v or v shr 1 # first round down to one less than a power of 2
+        v = v or v shr 2
+        v = v or v shr 4
+        v = v or v shr 8
+        v = v or v shr 16
+        v = v or v shr 32
+        result = lookup[(v * 0x03F6EAF2CD271461'u64) shr 58].int
+
+
+when not declared(leadingZeroBits) and declared(fastlog2):
+    proc leadingZeroBits*(x: SomeInteger): cint {.inline, nosideeffect.} =
+        ## Returns the number of leading zero bits in integer.
+        when sizeof(x) <= 4: result = sizeof(x)*8 - 1 - fastlog2(x.uint32).cint
+        else:                result = sizeof(x)*8 - 1 - fastlog2(x.uint64).cint
+
+when not declared(trailingZeroBits) and declared(firstSetBit):
+    proc trailingZeroBits*(x: SomeInteger): cint {.inline, nosideeffect.} =
+        ## Returns the number of trailing zeros in integer.
+        result = firstSetBit(x) - 1
+
+
+when not declared(parityBits):
+    # formula id from: https://graphics.stanford.edu/%7Eseander/bithacks.html#ParityParallel
+    template parity_impl[T](value: T): cint =
+        var v = value
+        when sizeof(T) == 8:
+            v = v xor (v shr 32)
+        when sizeof(T) >= 4:
+            v = v xor (v shr 16)
+        when sizeof(T) >= 2:
+            v = v xor (v shr 8)
+        v = v xor (v shr 4)
+        v = v and 0xf
+        ((0x6996'u shr v) and 1).cint
+
+    proc parityBits*(x: SomeInteger): cint {.inline, noSideEffect.} =
+        ## Calculate the bit parity in integer. If number of 1-bit
+        ## is odd parity is 1, otherwise 0.
+        when sizeof(x) <= 4: result = parity_impl(x.uint32)
+        else:                result = parity_impl(x.uint64)
+
+when not declared(countSetBits):
+    # generic formula is from: https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
+
+    proc countSetBits*(n: int8|uint8|int16|uint16|int32|uint32): cint {.inline, noSideEffect.} =
+      ## Counts the set bits in integer. (also called Hamming weight.)
+      var v = uint32(n)
+      v = v - ((v shr 1) and 0x55555555)
+      v = (v and 0x33333333) + ((v shr 2) and 0x33333333)
+      result = (((v + (v shr 4) and 0xF0F0F0F) * 0x1010101) shr 24).cint
+
+    proc countSetBits*(n: int|uint|int64|uint64): cint {.inline, noSideEffect.} =
+      ## Counts the set bits in integer. (also called Hamming weight.)
+      var v = uint64(n)
+      v = v - ((v shr 1'u64) and 0x5555555555555555'u64)
+      v = (v and 0x3333333333333333'u64) + ((v shr 2'u64) and 0x3333333333333333'u64)
+      v = (v + (v shr 4'u64) and 0x0F0F0F0F0F0F0F0F'u64)
+      result = ((v * 0x0101010101010101'u64) shr 56'u64).cint
+
+
+proc rotateLeftBits*(value: uint8;
+                     amount: static[range[1..7]]): uint8 {.inline, noSideEffect.} =
```

Comment:
Use 'range[1..7]' without the 'static' please, likewise for the following usages.